### DISCUSSION TITLE
Ender Pearl dupe glitch

### DISCUSSION THREAD

**Discussion by TheBusyBiscuit** [GriefPrevention/GriefPrevention#1307] [General]
Hi,

long time no see btw, how have you been? :)

Someone has recently reported a dupe glitch involving Ender Pearls through the use of Slimefun.
While I won't get into the details here on how this can be achieved to prevent abuse.
I managed to track it down to this exact line being run *relatively* unconditionally:

https://github.com/TechFortress/GriefPrevention/blob/746f10449d527410ae8af7815bdbd57671d449a4/src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java#L1097

The problem here being that some plugins, like Slimefun, allow Ender Pearls to be spawned out of thin air.
So I'd like your input on this on how to approach it the best way possible.
There are three potential solutions that would come to mind

a) **Metadata**
I could store metadata on the ender pearl which GP could listen to.
Only downside... from the `PlayerTeleportEvent` you cannot exactly obtain the enderpearl entity.

b) **Event**
There could be a `EnderPearlReturnEvent` added to GriefPrevention.
It is a bit of an awkward solution but would allow future integrations.

c) **TeleportCause**
I could change the TeleportCause on my end to `TeleportCause.CUSTOM`.
Though that is not quite inline with what is actually happening ingame since the teleportation will be triggered by an actual ender pearl and could have potential (unknown) side effects with other plugins.

I would love to hear your thoughts on how we could resolve this matter.
Let me know if you have any ideas or questions, thanks in advance!

**Comment by RoboMWM** [comment]
Yea glad to see the maven stuff is working out lol. My meme server where I use it has been kinda dormant for a while now so I haven't been watching it too much, but I do know it's almost always appearing in my auto-update scripts which was my primary motivation for mavenizing it lol. (But it did make it easier to contribute fixes/PRs too, I think one fix/PR I made to one of your projects I did blindly since I didn't have an easy way to compile and test it lol.)

Interesting.

I know with v20 (probably v19-ish timeframe) I plan to add events to GP when it decides to cancel stuff so you could handle it differently. (kinda like what you suggested for b). Only other idea I have is to somehow determine the "session" of a player throwing an enderpearl and return based on that I guess.

**Reply by TheBusyBiscuit** [reply]
> Yea glad to see the maven stuff is working out lol.

:D Yeah thanks for that, those PRs really gave me the kick I needed to finally come to turns with Maven. Nowadays I am quite experienced in it and I use it for every project, though it does have its downsides to it.

> I know with v20 (probably v19-ish timeframe) I plan to add events to GP when it decides to cancel stuff so you could handle it differently. (kinda like what you suggested for b). Only other idea I have is to somehow determine the "session" of a player throwing an enderpearl and return based on that I guess.

Hmmm, the event would definitely be an option. But the "session" (I assume you mean listening to players throwing an enderpearl and storing their uuid till the projectile hits) does sound more future-proof if similar dupes get discovered with other plugins.
How would you like to proceed? Since I don't know how you prefer dupes to be handled, I could send you a link to the issue or just tell them to re-open it on your repository. Or whether you would want to patch this secretly or want to go the integration-route with the event.

**Reply by RoboMWM** [reply]
Can open an issue here since I clearly forgot to get back to this lol
